cmd/vulcan-govulncheck/main.go

```diff
@@ -0,0 +1,208 @@
+		}
+	}
+
+	if opt.Depth == 0 {
+		opt.Depth = defaultDepth
+	}
+
+	logger.WithFields(logrus.Fields{"options": opt}).Debug("using options")
+
+	repoPath, _, err := helpers.CloneGitRepository(target, opt.Branch, opt.Depth)
+	if err != nil {
+		return fmt.Errorf("clone git repository: %w", err)
+	}
+
+	if err := runGovulncheck(ctx, repoPath, opt, state); err != nil {
+		return fmt.Errorf("run govulncheck: %w", err)
+	}
+
+	return nil
+}
+
+func runGovulncheck(ctx context.Context, dir string, opt options, state checkstate.State) error {
+	args := []string{"-format", "sarif"}
+	if opt.ChDir != "" {
+		args = append(args, "-C", opt.ChDir)
+	}
+	if opt.DB != "" {
+		args = append(args, "-db", opt.DB)
+	}
+	if opt.Mode != "" {
+		args = append(args, "-mode", opt.Mode)
+	}
+	if opt.Scan != "" {
+		args = append(args, "-scan", opt.Scan)
+	}
+	if opt.Tags != "" {
+		args = append(args, "-tags", opt.Tags)
+	}
+	if opt.Test != "" {
+		args = append(args, "-test", opt.Test)
+	}
+	if opt.Patterns != "" {
+		args = append(args, opt.Patterns)
+	}
+
+	var stdout, stderr bytes.Buffer
+	cmd := exec.CommandContext(ctx, "govulncheck", args...)
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	cmd.Dir = dir
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("run command: %w: %q", err, stderr.String())
+	}
+
+	vulns, err := parseSarif(&stdout)
+	if err != nil {
+		return fmt.Errorf("parse sarif: %w", err)
+	}
+
+	state.AddVulnerabilities(vulns...)
+	return nil
+}
+
+func parseSarif(r io.Reader) ([]report.Vulnerability, error) {
+	l, err := sarif.Decode(r)
+	if err != nil {
+		return nil, fmt.Errorf("decode SARIF document: %w", err)
+	}
+
+	var vulns []report.Vulnerability
+	for _, run := range l.Runs {
+		for _, result := range run.Results {
+			vuln, err := parseResult(l, result)
+			if err != nil {
+				return nil, fmt.Errorf("parse result: %w", err)
+			}
+			vulns = append(vulns, vuln)
+		}
+	}
+	return vulns, nil
+}
+
+func parseResult(l sarif.Log, result sarif.Result) (report.Vulnerability, error) {
+	rule, found := l.FindRule(result.RuleID)
+	if !found {
+		return report.Vulnerability{}, fmt.Errorf("unknown rule: %v", result.RuleID)
+	}
+
+	var rscs []report.ResourcesGroup
+	for _, cf := range result.CodeFlows {
+		rsc := report.ResourcesGroup{
+			Name:   cf.Message.Text,
+			Header: []string{"module", "location", "function"},
+		}
+		for _, tf := range cf.ThreadFlows {
+			for _, loc := range tf.Locations {
+				row := make(map[string]string)
+				row["module"] = loc.Module
+				row["location"] = loc.Location.PhysicalLocation.String()
+				row["function"] = loc.Location.Message.Text
+				rsc.Rows = append(rsc.Rows, row)
+			}
+		}
+		rscs = append(rscs, rsc)
+	}
+
+	location := getLocations(result)
+	vuln := report.Vulnerability{
+		Summary:          rule.ShortDescription.Text,
+		Score:            calcScore(result),
+		AffectedResource: location,
+		Fingerprint:      helpers.ComputeFingerprint(result.RuleID, location),
```

Comment:
The `location` is used to compute the fingerprint, which is fine. However, if there can be more than one location and the order is not deterministic, it would be beneficial to sort the slice to ensure the fingerprint is deterministic.

Reply:
I agree. We should be defensive here and do not rely on govulncheck's output to be always one location or multiple locations with the same order. Fixed.